src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,58 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
+            // according to ICU User Guide the performance of ucol_getSortKey is worse when it is called with null output buffer
+            // the solution is to try to fill the sort key in a temporary buffer of size equal 4 x string length
+
+            int guessedSize = (int)Math.Min((long)source.Length * 4, 1024 * 1024); // 1MB is the biggest array that can be rented from ArrayPool.Shared without memory allocation
+
+            if (TryGetSortKey(source, options, guessedSize, out int actualSortKeyLength, out int hash))
             {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+                return hash; // fast path, the buffer was big enough
+            }
 
-                byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
+            // the temporary buffer was too small but the actual sort key length was returned
+            if (actualSortKeyLength > 0 // 0 means internal error in ucol_getSortKey
+                && TryGetSortKey(source, options, actualSortKeyLength, out _, out hash)) 
+            {
+                return hash;
+            }
 
-                fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
-                {
-                    if (Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, sortKeyLength, options) != sortKeyLength)
-                    {
-                        throw new ArgumentException(SR.Arg_ExternalException);
-                    }
-                }
+            throw new ArgumentException(SR.Arg_ExternalException);
+        }
+        
+        private unsafe bool TryGetSortKey(ReadOnlySpan<char> source, CompareOptions options, int requestedSortKeyLength, out int actualSortKeyLength, out int hash)
+        {
+            Debug.Assert(requestedSortKeyLength > 0);
+            Debug.Assert(source.Length > 0);
 
-                int hash = Marvin.ComputeHash32(span.Slice(0, sortKeyLength), Marvin.DefaultSeed);
+            byte[]? borrowedArr = null;
+            Span<byte> sortKey = requestedSortKeyLength <= 512 ?
+                stackalloc byte[512] :
```

Comment:
Is it worth increasing to 1024? We use it elsewhere - eg., `Interop.Sys.StrError`, uses 1024 bytes, `IdnMapping.Unix.cs` uses 512 chars.

Reply:
ICU docs say "A recommended size for the temporary buffer is four times the length of the longest string processed." so I guess 512 bytes would be good for 64 chars which is not that long.

Reply:
The max string length is 128 (we got 512 bytes and the ICU describes it for unicode string length so 512 / 4 = 128). I have verified that.

But this is a Linux implementation which has a bigger stack size than Windows. @GrabYourPitchforks @jkotas would it be safe to increase the value to 1024 for Linux?

Reply:
Changing the stack buffer size to 1kB sounds find to me.

Reply:
I had not realized the default Linux stacks are larger. Looks like 8MB is common vs 1MB